DESIGN.md

```diff
@@ -61,8 +61,90 @@ specific authorizations
+
+Caveats describe which operations are authorized by providing predicates over
+the operation's attributes.
+
+Attributes are data, associated with the operation,
+that is known when the policy is evaluated, such as an identifier for the
+ressource being accessed, the type of the operation (read, write, append, ...),
+the operation's parameters (if any), the client's IP address or a
+channel-binding value (like the TLS transcript hash).
+
+Available attributes, and their type, are known ahead of time by the verifier.
+Some of those attributes are *critical*, and all caveats must provide a *bound*
+for each critical attribute.
+
+Bounds are a subset of predicates, that only allow the following:
+- `any`: all values match;
+- `in <subset>`: only elements in `subset` match; this can be an explicit
+  enumeration, or a (non-infinite) range in the case of numeric types.
+
+
+### Rationale
+
+Some attributes grant authority (such as ressource identifiers, operation type,
+...), and failing to include a caveat limiting acceptable values is a common
+failure with Macaroons, resulting in authority being accidentally granted.
+
+By marking them critical, two things are achieved:
+- They must be bound by caveats, preventing accidental authority grants when new
+  values are added.
+- Their presence is required in all caveats for a biscuit to be valid; as such:
+  - if developers accidentally fail to provide a bound, the biscuit is invalid;
+  - biscuits issued before the attribute was defined are implicitely revoked.
+
+For example, consider a data store, which initially only provides read access.
+Assume I was granted a biscuit for ressources in it, before a developper
+implemented read-write access, along with a `type` attribute (which can be
+`Read` or `Write`).  My biscuit suddenly grants me read-write access.
+
+Marking the `type` attribute as critical means that I must request a new
+biscuit, that properly specifies whether my access is read and/or write.
+
+Now, if I was to be issued a biscuit with the caveat `type != Write`, before the
+types `Append`, `Create`, and `Delete` were added, my the biscuit would again go
+from granting read-only access to granting write access; this is why critical
+attributes must use bounds.
+
+
+By requiring that all caveats provide a bound for each critical attribute, we
+can guarantee that a biscuit does not gain unintended authority when new
+attributes, or new values for them, are added in the system.  (The use of `any`
+is considered intentional.)
+
+
+### Interpretation
+
+Given an operation's `attributes`, the set of `critical` attributes, a given
+`biscuit` is evaluated as follows:
+
+```python3
+for caveat in biscuit:
+  bounds = set()
+  for predicate in caveat:
+    if not predicate.eval(attributes):
+      return False
+    if predicate.isbound:
+      bounds.add(predicate.attribute)
+
+  if bounds != critical:
+    return False
+
+return True
+```
+
+
 ## Format
 
+XXXTODO: Update for caveats
```

Comment:
Should we just remove the wire-format spec for now?

Reply:
I'd rather keep the format and the rights management part for now, but maybe adding a comment indicating we're working on those.

Reply:
OK; keeping as-is then (“XXXTODO: Update for [...]” should convey clearly that the section is out-of-date).